From 1a1677dcc86ffba45e8a7bb34d143d9fa73de466 Mon Sep 17 00:00:00 2001
From: Priyanka Bose <priyanka.bose@intel.com>
Date: Mon, 9 Jan 2023 07:12:42 +0000
Subject: [PATCH] PGO Optimization on CIV

Observed improved performance in launch time by 4-5% for apps related
to Google Maps, Photos, Kinemaster, Rar , Asphalt

Signed-off-by: Priyanka Bose <priyanka.bose@intel.com>
---
 runtime/Android.bp | 39 +++++++++++++++++++++++++++++++++++++++
 1 file changed, 39 insertions(+)

diff --git a/runtime/Android.bp b/runtime/Android.bp
index 1734b43eae..b31e6347ff 100644
--- a/runtime/Android.bp
+++ b/runtime/Android.bp
@@ -522,6 +522,36 @@ libart_static_cc_defaults {
     },
 }
 
+libart_cc_defaults {
+     name: "libart-pgo-defaults",
+     pgo: {
+         instrumentation: true,
+         benchmarks: ["art_runtime", "dex2oat",],
+     },
+     target: {
+         android_arm64: {
+             pgo: {
+                 profile_file: "art/art-runtime_arm_arm64.profdata",
+             },
+         },
+         android_arm: {
+             pgo: {
+                 profile_file: "art/art-runtime_arm_arm64.profdata",
+             },
+         },
+         android_x86_64: {
+             pgo: {
+                 profile_file: "art/art-runtime-x86-x86_64.profdata.2023-01-09",
+             },
+         },
+         android_x86: {
+             pgo: {
+                 profile_file: "art/art-runtime-x86-x86_64.profdata.2023-01-09",
+             },
+         },
+     },
+}
+
 cc_defaults {
     name: "libart_static_defaults",
     defaults: [
@@ -598,6 +628,7 @@ art_cc_defaults {
         "libart_defaults",
         "libart_nativeunwind_defaults",
         "art_hugepage_defaults",
+        "libart-pgo-defaults",
     ],
     whole_static_libs: [
     ],
@@ -612,6 +643,14 @@ art_cc_defaults {
     export_shared_lib_headers: [
         "libdexfile",
     ],
+    pgo: {
+         // Additional cflags just for dex2oat during PGO instrumentation
+         cflags: [
+               // Ignore frame-size increase resulting from instrumentation.
+               "-Wno-frame-larger-than=",
+               "-DART_PGO_INSTRUMENTATION",
+      ],
+    },    
     target: {
         android: {
             lto: {
-- 
2.37.3

